Uniformly apply Windows logic in cpp_extensions everywhere

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31162 Don't use RTLD_GLOBAL to load _C.
* **#31161 Uniformly apply Windows logic in cpp_extensions everywhere**

Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

ezyang committed:
Actually get rid of Windows hacks in cpp_extensions.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit a12a58a:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a12a58a8424f25610721f9e241c699f976a38e60).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_macos_10_13_py3_test</sup> | <sup>Test</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Actually get rid of Windows hacks in cpp_extensions."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Actually get rid of Windows hacks in cpp_extensions."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Actually get rid of Windows hacks in cpp_extensions."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Actually get rid of Windows hacks in cpp_extensions."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Actually get rid of Windows hacks in cpp_extensions."


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]

soumith reviewed (approved):
lgtm

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]

yf225 commented:
@peterjc123 Would you like to take a look?

peterjc123 commented on torch/utils/cpp_extension.py:
```diff
@@ -497,14 +495,12 @@ def CUDAExtension(name, sources, *args, **kwargs):
-        libraries.append('_C')
+    libraries.append('c10')
+    libraries.append('c10_cuda')
+    libraries.append('torch')
```
What is `torch` here used for? The functions should be in `torch_cpu` and `torch_cuda`, right?

ezyang replied:
Yeah, it's supposed to be an empty library. I put `torch` in here for "good luck", in case someone ever accidentally puts some symbols in it.

peterjc123 commented on torch/utils/cpp_extension.py:
```diff
@@ -446,18 +446,16 @@ def CppExtension(name, sources, *args, **kwargs):
+
+    libraries = kwargs.get('libraries', [])
+    libraries.append('c10')
+    libraries.append('torch')
```
and here?

peterjc123 commented on torch/utils/cpp_extension.py:
```diff
@@ -949,9 +944,18 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
+        extra_ldflags.append('-ltorch_cpu')
+        if with_cuda:
+            extra_ldflags.append('-ltorch_cuda')
+        extra_ldflags.append('-ltorch')
```
and here?

ezyang committed:
Update on "Uniformly apply Windows logic in cpp_extensions everywhere"


Previously, it wasn't necessary to specify `DT_NEEDED` in C++ extensions on Linux (aka pass `-l` flags) because all of the symbols would have already been loaded with `RTLD_GLOBAL`, so there wouldn't be any undefined symbols.  But when we switch to loading `_C` with `RTLD_LOCAL`, it's now necessary for all the C++ extensions to know what libraries to link with. The resulting code is clearer and more uniform, so it's wins all around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262578](https://our.internmc.facebook.com/intern/diff/D19262578)

[ghstack-poisoned]
